presto-hive/src/main/java/com/facebook/presto/hive/HiveStagingFileCommitter.java

```diff
@@ -0,0 +1,71 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.HdfsEnvironment.HdfsContext;
+import com.facebook.presto.hive.PartitionUpdate.FileWriteInfo;
+import com.facebook.presto.spi.ConnectorSession;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import javax.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.presto.hive.metastore.MetastoreUtil.getFileSystem;
+import static com.facebook.presto.hive.metastore.MetastoreUtil.renameFile;
+import static com.facebook.presto.hive.metastore.MetastoreUtil.waitForListenableFutures;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+
+public class HiveStagingFileCommitter
+        implements StagingFileCommitter
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final ListeningExecutorService commitExecutor;
+
+    @Inject
+    public HiveStagingFileCommitter(
+            HdfsEnvironment hdfsEnvironment,
+            @ForFileRename ExecutorService executorService)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.commitExecutor = listeningDecorator(executorService);
+    }
+
+    @Override
+    public void commitFiles(ConnectorSession session, String schemaName, String tableName, List<PartitionUpdate> partitionUpdates)
+    {
+        HdfsContext context = new HdfsContext(session, schemaName, tableName);
+        List<ListenableFuture<?>> commitFutures = new ArrayList<>();
+
+        for (PartitionUpdate partitionUpdate : partitionUpdates) {
+            Path path = partitionUpdate.getWritePath();
+            FileSystem fileSystem = getFileSystem(hdfsEnvironment, context, path);
+            for (FileWriteInfo fileWriteInfo : partitionUpdate.getFileWriteInfos()) {
+                checkState(!fileWriteInfo.getWriteFileName().equals(fileWriteInfo.getTargetFileName()));
+                Path source = new Path(path, fileWriteInfo.getWriteFileName());
+                Path target = new Path(path, fileWriteInfo.getTargetFileName());
+                commitFutures.add(commitExecutor.submit(() -> renameFile(fileSystem, source, target)));
+            }
+        }
+
+        waitForListenableFutures(commitFutures);
```

Comment:
Just some thoughts.

I can see how the `waitForListenableFutures` saves the code lines, and follows the DRY principles. However the semantics of this method are very unclear. It is very hard to know what does this method do. Without looking into the method guts it is impossible to give answers to these questions:

- Single future has failed? Does it fail? Does it fail fast?
- Does it cancel the features in case of a failure?
- How are the failures propagated?

I don't have a strong opinion here. But it seems to me that in such tricky cases sometimes it is better to leave a few lines more, but use the standard library methods (that engineers are more likely to know semantics of).

But again, no action is required here. Those are just thoughts.

Reply:
I agree with @arhimondr . Especially given it's only used twice, inline it seems to be reasonable.

While DRY principal is in general a good practice, there are cases it doesn't apply. It really depends on:
- Does the refactor/abstraction natural ? -- does it make things difficult to follow ?
- How evil is to repeat your-self in this case? 
- Are these repeating real semantic repeat, or it's just the code look very similar?